core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/workflow/WorkflowVersionResource.scala

```diff
@@ -30,7 +33,98 @@ import scala.collection.convert.ImplicitConversions.`collection AsScalaIterable`
 object WorkflowVersionResource {
   final private lazy val context = SqlServer.createDSLContext()
   final private val workflowDao = new WorkflowDao(context.configuration)
+  // constant to indicate versions should be aggregated if they are within the specified time limit
+  private final val AGGREGATE_TIME_LIMIT_MILLSEC =
+    AmberUtils.amberConfig.getInt("user-sys.version-time-limit") * 60000
+  // list of Json keys in the diff patch that are considered UNimportant
+  private final val VERSION_UNIMPORTANCE_RULES = List("/operatorPositions/")
 
+  /**
+    * This function gives a label to each version whether it is significant or not based on a few rules
+    * The reason why it is computed AFTER retrieving the list of versions is due to multiple reasons:
+    * * 1. minimize the changes to the database.
+    * * 2. since the frontend sends persisting workflow request often, we don't want to slow down the
+    * insertion to DB because of computing the version's importance especially because the request is
+    * async, the versions can quickly become inconsistent if there is delay.
+    * * 3. The rules can be changed in the future so we want this logic to be changed flexibly.
+    * @param versions
+    * @return
+    */
+  private def encodeVersionImportance(
+      versions: List[VersionEntry]
```

Comment:
I suggest make this api take one version only, then use `versions.map(encodeVersionImportance)` to make the api more flexible and easier to write (for example you do not need empty list check). minor suggestion.

unless in the future, version's importance might depend on the order.

Reply:
Yes I am depending on the timestamps of the versions in the ordered list to determine if they should be collapsed under each other or not.